syntax = "proto3";
package users;
option go_package = "users";


message User {
	optional uint32 id = 1;
	optional string name = 2;
	optional string email = 3;
	optional string password = 4;
	optional string created_at = 5;
	optional string updated_at = 6;
}


message GetUserRequest {
  	uint32 id = 1;
}
message GetUserResponse {
	User user = 1;
}

message CreateUserRequest {
	User user = 1;
}
message CreateUserResponse {
	string message = 1;
}

message UpdateUserRequest {
	uint32 id = 1;
	User user = 2;
}
message UpdateUserResponse {
	User user = 1;
}

message DeleteUserRequest {
	uint32 id = 1;
}
message DeleteUserResponse {
	bool status = 1;
}

message ListUsersRequest {
	optional string filter = 1;
	optional string order_direction = 2;
	optional string order_by = 3;
	optional uint32 page = 4;
	optional uint32 limit = 5;
}
message Meta {
	uint32 total = 1;
	uint32 page = 2;
	uint32 limit = 3;
} 
message ListUsersResponse {
	repeated User users = 1;
	Meta meta = 2;
}

message LoginUserRequest {
	string email = 1;
	string password = 2;
}
message LoginUserResponse {
	string token = 1;
	bool status = 2;
}


service UserService {	
	rpc GetUser(GetUserRequest) returns (GetUserResponse);
	rpc CreateUser(CreateUserRequest) returns (CreateUserResponse);
	rpc UpdateUser(UpdateUserRequest) returns (UpdateUserResponse);
	rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);
	rpc ListUsers(ListUsersRequest) returns (ListUsersResponse);
	rpc LoginUser(LoginUserRequest) returns (LoginUserResponse);
}

